Problem with generics

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: windows, Microsoft Windows 10 Pro v19044 64 bits
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i5-3570 CPU @ 3.40GHz  

CC version: Error: exec failed (CreateProcess) with code 2: O sistema não pode encontrar o arquivo especificado.

 cmd: cc --version

getwd: C:\dev\testv\bug_02
vmodules: C:\Users\Nereu\.vmodules
vroot: C:\dev\v
vexe: C:\dev\v\v.exe
vexe mtime: 2022-10-04 16:56:30
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.1 6ac9552.acbfe88

Git version: git version 2.29.1.windows.1
Git vroot status: weekly.2022.40-2-gacbfe88d
.git/config present: true
thirdparty/tcc status: thirdparty-windows-amd64 125e21e2

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg .\main.v`
```v
fn sum<T>(a T,b T) T {
	return a + b
}

fn main(){
  sum({"id":1},{"id":1})
}
```

**What did you expect to see?**

error message


**What did you see instead?**
```
==================
C:/Users/Nereu/AppData/Local/Temp/v_0/main.1016228435805903514.tmp.c:7390: warning: cast between pointer and integer of different size
C:/Users/Nereu/AppData/Local/Temp/v_0/main.1016228435805903514.tmp.c:12194: error: invalid operand types for binary operation
C:/Users/Nereu/AppData/Local/Temp/v_0/main.1016228435805903514.tmp.c:12194: error: invalid aggregate type for register load
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/checker/infix.v
+++ b/vlib/v/checker/infix.v
@@ -164,19 +164,24 @@ pub fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 		}
 		.plus, .minus, .mul, .div, .mod, .xor, .amp, .pipe {
 			// binary operators that expect matching types
+			unwrapped_left_type := c.unwrap_generic(left_type)
+			left_sym = c.table.sym(unwrapped_left_type)
+			unwrapped_right_type := c.unwrap_generic(right_type)
+			right_sym = c.table.sym(unwrapped_right_type)
 			if right_sym.info is ast.Alias && (right_sym.info as ast.Alias).language != .c
-				&& c.mod == c.table.type_to_str(right_type).split('.')[0]
+				&& c.mod == c.table.type_to_str(unwrapped_right_type).split('.')[0]
 				&& c.table.sym((right_sym.info as ast.Alias).parent_type).is_primitive() {
 				right_sym = c.table.sym((right_sym.info as ast.Alias).parent_type)
 			}
 			if left_sym.info is ast.Alias && (left_sym.info as ast.Alias).language != .c
-				&& c.mod == c.table.type_to_str(left_type).split('.')[0]
+				&& c.mod == c.table.type_to_str(unwrapped_left_type).split('.')[0]
 				&& c.table.sym((left_sym.info as ast.Alias).parent_type).is_primitive() {
 				left_sym = c.table.sym((left_sym.info as ast.Alias).parent_type)
 			}
 
 			if c.pref.translated && node.op in [.plus, .minus, .mul]
-				&& left_type.is_any_kind_of_pointer() && right_type.is_any_kind_of_pointer() {
+				&& unwrapped_left_type.is_any_kind_of_pointer()
+				&& unwrapped_right_type.is_any_kind_of_pointer() {
 				return_type = left_type
 			} else if !c.pref.translated && left_sym.kind == .alias && left_sym.info is ast.Alias
 				&& !(c.table.sym((left_sym.info as ast.Alias).parent_type).is_primitive()) {
@@ -193,8 +198,8 @@ pub fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 						return_type = left_type
 					}
 				} else {
-					left_name := c.table.type_to_str(left_type)
-					right_name := c.table.type_to_str(right_type)
+					left_name := c.table.type_to_str(unwrapped_left_type)
+					right_name := c.table.type_to_str(unwrapped_right_type)
 					if left_name == right_name {
 						c.error('undefined operation `$left_name` $node.op.str() `$right_name`',
 							left_right_pos)
@@ -217,8 +222,8 @@ pub fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 						return_type = right_type
 					}
 				} else {
-					left_name := c.table.type_to_str(left_type)
-					right_name := c.table.type_to_str(right_type)
+					left_name := c.table.type_to_str(unwrapped_left_type)
+					right_name := c.table.type_to_str(unwrapped_right_type)
 					if left_name == right_name {
 						c.error('undefined operation `$left_name` $node.op.str() `$right_name`',
 							left_right_pos)
@@ -236,8 +241,8 @@ pub fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 						return_type = left_type
 					}
 				} else {
-					left_name := c.table.type_to_str(left_type)
-					right_name := c.table.type_to_str(right_type)
+					left_name := c.table.type_to_str(unwrapped_left_type)
+					right_name := c.table.type_to_str(unwrapped_right_type)
 					if left_name == right_name {
 						c.error('undefined operation `$left_name` $node.op.str() `$right_name`',
 							left_right_pos)
@@ -253,8 +258,8 @@ pub fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 						return_type = right_type
 					}
 				} else {
-					left_name := c.table.type_to_str(left_type)
-					right_name := c.table.type_to_str(right_type)
+					left_name := c.table.type_to_str(unwrapped_left_type)
+					right_name := c.table.type_to_str(unwrapped_right_type)
 					if left_name == right_name {
 						c.error('undefined operation `$left_name` $node.op.str() `$right_name`',
 							left_right_pos)
@@ -264,14 +269,14 @@ pub fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 				}
 			} else if node.left.is_auto_deref_var() || node.right.is_auto_deref_var() {
 				deref_left_type := if node.left.is_auto_deref_var() {
-					left_type.deref()
+					unwrapped_left_type.deref()
 				} else {
-					left_type
+					unwrapped_left_type
 				}
 				deref_right_type := if node.right.is_auto_deref_var() {
-					right_type.deref()
+					unwrapped_right_type.deref()
 				} else {
-					right_type
+					unwrapped_right_type
 				}
 				left_name := c.table.type_to_str(ast.mktyp(deref_left_type))
 				right_name := c.table.type_to_str(ast.mktyp(deref_right_type))
@@ -279,8 +284,8 @@ pub fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 					c.error('mismatched types `$left_name` and `$right_name`', left_right_pos)
 				}
 			} else {
-				unaliased_left_type := c.table.unalias_num_type(left_type)
-				unalias_right_type := c.table.unalias_num_type(right_type)
+				unaliased_left_type := c.table.unalias_num_type(unwrapped_left_type)
+				unalias_right_type := c.table.unalias_num_type(unwrapped_right_type)
 				mut promoted_type := c.promote_keeping_aliases(unaliased_left_type, unalias_right_type,
 					left_sym.kind, right_sym.kind)
 				// substract pointers is allowed in unsafe block
@@ -290,17 +295,21 @@ pub fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 					promoted_type = ast.int_type
 				}
 				if promoted_type.idx() == ast.void_type_idx {
-					left_name := c.table.type_to_str(left_type)
-					right_name := c.table.type_to_str(right_type)
+					left_name := c.table.type_to_str(unwrapped_left_type)
+					right_name := c.table.type_to_str(unwrapped_right_type)
 					c.error('mismatched types `$left_name` and `$right_name`', left_right_pos)
 				} else if promoted_type.has_flag(.optional) {
 					s := c.table.type_to_str(promoted_type)
 					c.error('`$node.op` cannot be used with `$s`', node.pos)
 				} else if promoted_type.is_float() {
 					if node.op in [.mod, .xor, .amp, .pipe] {
-						side := if left_type == promoted_type { 'left' } else { 'right' }
-						pos := if left_type == promoted_type { left_pos } else { right_pos }
-						name := if left_type == promoted_type {
+						side := if unwrapped_left_type == promoted_type { 'left' } else { 'right' }
+						pos := if unwrapped_left_type == promoted_type {
+							left_pos
+						} else {
+							right_pos
+						}
+						name := if unwrapped_left_type == promoted_type {
 							left_sym.name
 						} else {
 							right_sym.name

--- a/vlib/v/checker/tests/generic_infix_plus_err.out
+++ b/vlib/v/checker/tests/generic_infix_plus_err.out
@@ -1,0 +1,6 @@
+vlib/v/checker/tests/generic_infix_plus_err.vv:2:9: error: undefined operation `map[string]int` + `map[string]int`
+    1 | fn sum<T>(a T, b T) T {
+    2 |     return a + b
+      |            ~~~~~
+    3 | }
+    4 |

--- a/vlib/v/checker/tests/generic_infix_plus_err.vv
+++ b/vlib/v/checker/tests/generic_infix_plus_err.vv
@@ -1,0 +1,11 @@
+fn sum<T>(a T, b T) T {
+	return a + b
+}
+
+fn main() {
+	sum({
+		'id': 1
+	}, {
+		'id': 1
+	})
+}

